ui/core/player.ts

```diff
@@ -1138,17 +1159,22 @@ export class Player<SpecType extends Spec> {
 
 	computeItemEP(item: Item, slot: ItemSlot): number {
 		if (item == null) return 0;
+		const cacheKey = `${item.id}-${this.challengeModeEnabled}`;
 
-		const cached = this.itemEPCache[slot].get(item.id);
+		const cached = this.itemEPCache[slot].get(cacheKey);
 		if (cached !== undefined) return cached;
 
-		const itemStats = new Stats(item.stats).add(getWeaponStatsBySlot(item, slot));
+		const equippedItem = new EquippedItem({
+			item,
+			challengeMode: this.challengeModeEnabled,
+		}).getWithDynamicStats();
+		const itemStats = new Stats(equippedItem._item.stats).add(getWeaponStatsBySlot(item, slot, equippedItem.upgrade));
```

Comment:
It might be cleaner to combine these three calls into a single instance method `calcStats(slot: ItemSlot): Stats` on `EquippedItem`. That way you don't need to clone a whole other `EquippedItem` object with the `UIItem` overwritten, and can replace `getWithDynamicStats()` with a direct computation of the desired `Stats` object containing both the raw stats and the weapon PseudoStats.